### PR TITLE
Add typing indicators dialogues

### PR DESCRIPTION
**Pull request by lagerros**
UPDATED: 

Hi all, 

This PR adds some text that says how many other users are currently typing in a dialogue (technically: typed within the last 20 seconds). 

* Added relevant types needed by both server and client
* Adding a DebateTypingIndicator component (new file), and displaying it on DebateBody component
* Made some type changes to the useUnreadNotifications hook
* Created a new collection TypingIndicators along with schema and fragment
* Added a resolver to handle the server response
* Added a TypingIndicators repo
* Added a check in serverSentEvents.ts that runs every second to check for new typing indicators

image
image
image

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1205484537706250) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/server/serverSentEvents.ts [hunk 1]
```diff
@@ -14,6 +17,7 @@ interface ConnectionInfo {
 }
 const openConnections: Record<string, ConnectionInfo[]> = {};
 
+// TODO: rename this from /notificationEvents to something more generic
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:20]: Probably not worth it, changing endpoints is a pain (and has pretty minimal benefits; "notification" is sufficiently ambiguous that the name isn't even wrong when used for this expanded purposes)

File: packages/lesswrong/server/serverSentEvents.ts [hunk 2]
```diff
@@ -71,10 +75,137 @@ export function addServerSentEventsEndpoint(app: Express) {
...
+  const listOfPosts = uniq(newTypingIndicatorsDummy.map(indicator => indicator.documentId))
+
+  const posts = await Posts.find({
+    _id: {$in:listOfPosts}}, {
+    projection: {_id: 1, userId: 1, coauthorStatuses: 1, shareWithUsers: 1}
+  }).fetch()
+
+  const postsWithUserIds: Record<string, string[]> = {}
+  for (let post of posts) {
+    const coauthorIds = post.coauthorStatuses ? post.coauthorStatuses.map(coauthor => coauthor.userId) : []
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:156]: For consistency's sake, probably best to use the `getConfirmedCoauthorIds` function here instead

File: packages/lesswrong/server/serverSentEvents.ts [hunk 3]
```diff
@@ -71,10 +75,137 @@ export function addServerSentEventsEndpoint(app: Express) {
...
+    const shareWithUsersIds = post.shareWithUsers ? post.shareWithUsers : []
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:157]: If `shareWithUsers` can be `null`, you can just do `const shareWithUsersIds = post.shareWithUsers ?? []`

File: packages/lesswrong/server/serverSentEvents.ts [hunk 4]
```diff
@@ -71,10 +75,137 @@ export function addServerSentEventsEndpoint(app: Express) {
...
+      if (userId in usersWithPostIds) {
+        usersWithPostIds[userId].push(postId)
+      } else {
+        usersWithPostIds[userId] = [postId]
+      }
+    }
+  }
+
+  const postsWithTypingIndicators: Record<string, TypingIndicatorInfo[]> = {}
+  for (let typingIndicator of newTypingIndicatorsDummy) {
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:175]: Don't forget to filter these for `collectionName: 'Posts'` (assuming the non-post ones, should they ever exist, will be supported elsewhere)

File: packages/lesswrong/server/serverSentEvents.ts [hunk 5]
```diff
@@ -71,10 +75,137 @@ export function addServerSentEventsEndpoint(app: Express) {
...
+  for (let postId of Object.keys(postsWithUserIds)) {
+    const userIds = postsWithUserIds[postId]
+    for (let userId of userIds) {
+      if (userId in usersWithPostIds) {
+        usersWithPostIds[userId].push(postId)
+      } else {
+        usersWithPostIds[userId] = [postId]
+      }
+    }
+  }
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:172]: May be redundant given https://github.com/ForumMagnum/ForumMagnum/pull/7857/files#r1323832568, but I don't love the repetitive & difficult-to-read structure of the logic that collects `usersWithPostIds`, `postsWithTypingIndicators`, and `usersReceivingTypingIndicators`; they all feel like they could just be pretty straightforward `.map` calls instead.

File: packages/lesswrong/server/serverSentEvents.ts (hunk 4 above)
Comment (lagerros) [reply packages/lesswrong/server/serverSentEvents.ts:175]: ignoring per conversatoin in person

File: packages/lesswrong/lib/collections/typingIndicators/collection.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+export const TypingIndicators: TypingIndicatorsCollection = createCollection({
+  collectionName: 'TypingIndicators',
+  typeName: 'TypingIndicator',
+  collectionType: 'pg',
+  schema,
+  resolvers: getDefaultResolvers('TypingIndicators'),
+  mutations: getDefaultMutations('TypingIndicators', options),
+  logChanges: true,
+})
+
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/typingIndicators/collection.ts:33]: I think you also want to define a `TypingIndicators.checkAccess` which default-denies access via our GraphQL API, since legit access will only be going through direct DB query inside the SSE logic

File: packages/lesswrong/server/serverSentEvents.ts [hunk 7]
```diff
@@ -5,6 +5,9 @@ import { getSiteUrl } from "../lib/vulcan-lib/utils";
 import { DatabaseServerSetting } from './databaseSettings';
 import maxBy from 'lodash/maxBy';
 import moment from 'moment';
+import { Posts } from '../lib/collections/posts';
+import uniq from 'lodash/uniq';
+import { ConnectionMessage } from '../client/serverSentEventsClient';
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:10]: `import type`

File: packages/lesswrong/server/serverSentEvents.ts [hunk 8]
```diff
@@ -71,10 +75,137 @@ export function addServerSentEventsEndpoint(app: Express) {
...
+  //dialog on stuff has 
+    // raemon, testcob ultron
+  // coconut has
+    // jacob, raemontest
+  // Otter world
+    // jacobjacob, kave, habryka, raemon
+
+  const listOfPosts = uniq(newTypingIndicatorsDummy.map(indicator => indicator.documentId))
+
+  const posts = await Posts.find({
```
[the remark this answers is not in the document]
Comment (lagerros) [reply packages/lesswrong/server/serverSentEvents.ts:149]: tracking

File: packages/lesswrong/server/serverSentEvents.ts (hunk 5 above)
Comment (lagerros) [reply packages/lesswrong/server/serverSentEvents.ts:172]: tracking

File: packages/lesswrong/components/comments/DebateTypingIndicator.tsx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    ${fragmentTextForQuery("TypingIndicatorsDefaultFragment")}
+  `)
+
+  useGlobalKeydown(throttle(() => {
+    void upsertTypingIndicator({variables: {documentId: post._id}})
+  }, 300));
+
+  useOnNotificationsChanged(currentUser, (messageString) => {
+    const message : TypingIndicatorMessage = JSON.parse(messageString)
```
Comment (b0b3rt) [inline packages/lesswrong/components/comments/DebateTypingIndicator.tsx:39]: note: directly assigning the output of `JSON.parse` to something isn't super safe; generally you'd want to validate that it's actually the thing you're expecting.  In this case I think it's actually a `ConnectionMessage` rather than a `TypingIndicatorMessage`, and you need to check what the `eventType` value is before using it?  (And if it's a `NotificationCheckMessage`, you don't actually want to execute any logic on it in this component.)

File: packages/lesswrong/components/hooks/useUnreadNotifications.ts [hunk 10]
```diff
@@ -69,8 +69,10 @@ export function useUnreadNotifications(): {
   const unreadPrivateMessages = data?.unreadNotificationCounts?.unreadPrivateMessages ?? 0;
   const checkedAt = data?.unreadNotificationCounts?.checkedAt || null;
 
-  const refetchIfNewNotifications = useCallback((timestamp: Date) => {
-    if (!checkedAt || timestamp > checkedAt) {
+  const refetchIfNewNotifications = useCallback((messageBlob: string) => {
+    // TODO: figure out what's going on with messageBlob (which was formerly timestamp, but,
```
Comment (b0b3rt) [inline packages/lesswrong/components/hooks/useUnreadNotifications.ts:73]: Pretty sure Jim's PR here has a fix for some incorrect typing going on: https://github.com/ForumMagnum/ForumMagnum/pull/7828

File: packages/lesswrong/lib/collections/typingIndicators/schema.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  userId: {
+    type: String,
+    optional: true,
+    nullable: false,
+    canRead: ['members'],
+    canCreate: ['members'],
+  },
+  documentId: {
+    type: String,
+    optional: true,
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/typingIndicators/schema.ts:13]: Remove `optional: true` for these; that's a field that's only relevant for generating forms and we aren't generating forms based on TypingIndicators

File: packages/lesswrong/lib/collections/typingIndicators/views.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+  }
+}
+
+TypingIndicators.addView("typingIndicatorsForPost", function (terms: TypingIndicatorsViewTerms) {
+  return {
+    selector: {documentId: terms.documentId},
+    options: {limit: 1}
+  }
+})
+ensureIndex(TypingIndicators, { documentId: 1 })
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/typingIndicators/views.ts:17]: Remove this index, it's mostly redundant with the one [here](https://github.com/ForumMagnum/ForumMagnum/pull/7857/files#diff-e0ab4ceed1c2f1705c22eed975841c262036e75021df3b99d43427304e6b3a1bR42)

File: packages/lesswrong/lib/sql/PgCollection.ts [hunk 13]
```diff
@@ -222,8 +222,14 @@ class PgCollection<T extends DbObject> extends MongoCollection<T> {
...
       ? {[fieldOrSpec as keyof T]: 1 as const} as MongoIndexKeyObj<T>
       : fieldOrSpec;
+    
     const index = this.table.getIndex(Object.keys(key), options) ?? this.getTable().addIndex(key, options);
+    
     const query = new CreateIndexQuery(this.getTable(), index, true);
+    if (this.collectionName === "TypingIndicators") {
+      console.log(query, fieldOrSpec)
+    }
+
```
Comment (b0b3rt) [inline packages/lesswrong/lib/sql/PgCollection.ts:232]: Don't forget to undo these changes

File: packages/lesswrong/server/resolvers/typingIndicatorsResolvers.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import {getConfirmedCoauthorIds} from "../../lib/collections/posts/helpers";
+import TypingIndicatorsRepo from "../repos/TypingIndicatorsRepo";
+import {defineMutation} from "../utils/serverGraphqlUtil";
+
+function isUserDialogueParticipant(userId: string, post: DbPost) {
+  // if  (!post.debate) return false
```
Comment (b0b3rt) [inline packages/lesswrong/server/resolvers/typingIndicatorsResolvers.ts:6]: clean up

File: packages/lesswrong/server/resolvers/typingIndicatorsResolvers.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  fn: async (_, {documentId}:{documentId:string}, {currentUser, loaders}) => {
+    if (!currentUser) throw new Error("No user was provided")
+    const post = await loaders.Posts.load(documentId)
+    if (!post) throw new Error("No post was provided")
+    if (!post.debate) throw new Error("Post is not a dialogue")
+    if (!isUserDialogueParticipant(currentUser._id, post)) throw new Error("User is not a dialog participant")
+
+    await new TypingIndicatorsRepo().upsertTypingIndicator(currentUser._id, post._id)
+  } 
+})
```
Comment (b0b3rt) [inline packages/lesswrong/server/resolvers/typingIndicatorsResolvers.ts:26]: newline at end of file

File: packages/lesswrong/server/serverSentEvents.ts [hunk 16]
```diff
@@ -129,7 +263,7 @@ async function checkForNotifications() {
         if (!connection.newestNotificationTimestamp
           || connection.newestNotificationTimestamp < newTimestamp
         ) {
-          const message = { newestNotificationTime: newTimestamp };
+          const message: ConnectionMessage = { eventType: "notificationCheck", newestNotificationTime: newTimestamp };
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:214]: this type should be the narrower `NotificationCheckMessage`

File: packages/lesswrong/server/serverSentEvents.ts [hunk 17]
```diff
@@ -71,10 +76,139 @@ export function addServerSentEventsEndpoint(app: Express) {
...
+      }
+    }
+  }
+
+  console.log(usersReceivingTypingIndicators)
+
+  for (let userId of Object.keys(usersReceivingTypingIndicators)) {
+    if (openConnections[userId]) {
+      for (let connection of openConnections[userId]) {
+        const message : ConnectionMessage = {eventType: "typingIndicator", typingIndicators: usersReceivingTypingIndicators[userId]}
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:152]: this type should be the narrower `TypingIndicatorMessage`

File: packages/lesswrong/lib/collections/typingIndicators/collection.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+  editCheck: (user: DbUser|null, document: DbTypingIndicator|null) => {
+    if (!user || !document) return false;
+    return userOwns(user, document) 
+  },
+
+  removeCheck: (user: DbUser|null, document: DbTypingIndicator|null) => {
+    if (!user || !document) return false;
+    return userOwns(user, document) 
+  },
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/typingIndicators/collection.ts:22]: I think.. we should probably not grant people the ability to delete typing indicators, even if they own them, and to the extent that we allow this for other collections it's a legacy decision I'd want to move away from

File: packages/lesswrong/components/comments/DebateTypingIndicator.tsx [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+    ${fragmentTextForQuery("TypingIndicatorsDefaultFragment")}
+  `)
+
+  useGlobalKeydown(throttle(() => {
+    void upsertTypingIndicator({variables: {documentId: post._id}})
+  }, 300));
+
+  useOnNotificationsChanged(currentUser, (message) => {
+    if (message.eventType === 'typingIndicator') {
+      const typingIndicators = message.typingIndicators ?? []
```
Comment (b0b3rt) [inline packages/lesswrong/components/comments/DebateTypingIndicator.tsx:38]: I don't think `?? []` is necessary

File: packages/lesswrong/server/serverSentEvents.ts [hunk 20]
```diff
@@ -72,9 +77,65 @@ export function addServerSentEventsEndpoint(app: Express) {
...
+        prev[userIdToNotify] = [{_id, userId, documentId, lastUpdated}]
+      }
+    })
+    return prev
+  }, results)
+
+  for (let userId of Object.keys(results)) {
+    if (openConnections[userId]) {
+      for (let connection of openConnections[userId]) {
+        const message : ServerSentEventsMessage = {eventType: "typingIndicator", typingIndicators: results[userId]}
```
Comment (b0b3rt) [inline packages/lesswrong/server/serverSentEvents.ts:133]: This should be the narrower type specific to typing indicators

File: packages/lesswrong/components/comments/DebateTypingIndicator.tsx [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+      })
+      setTypingIndicators(filteredIndicators)
+    }
+  });
+
+  if (!currentUser) return null;
+
+  const otherUsers = typingIndicators.filter((typingIndicator) => {
+    const twentySecondsAgo = Date.now() - 20000
+    const typingIndicatorIsRecent = (new Date(typingIndicator.lastUpdated ?? 0).getTime()) > twentySecondsAgo;
```
Comment (b0b3rt) [inline packages/lesswrong/components/comments/DebateTypingIndicator.tsx:50]: I don't think `typingIndicator.lastUpdated` will ever be null, so the coalescing shouldn't be necessary

**Review by b0b3rt — APPROVED** [review]
lgtm, pending testing
